File: salazar/salazar/agents/generator.py
"""Generator agent — builds one feature per session."""

import asyncio
import json
import logging
from pathlib import Path

from claude_agent_sdk import query, ResultMessage

from salazar.client import make_options, OUTPUT_DIR
from salazar.progress import read_progress, format_progress_header

logger = logging.getLogger("harness.generator")

PROMPT_PATH = Path(__file__).resolve().parent.parent / "prompts" / "generator.md"


async def run_generator(feature: dict, evaluator_feedback: str | None = None) -> dict:
    """Run the generator agent to implement a single feature.

    Args:
        feature: The feature dict from feature_list.json.
        evaluator_feedback: If retrying, the evaluator's feedback from the previous attempt.

    Returns:
        Dict with: success (bool), cost_usd (float), error (str|None)
    """
    feature_id = feature.get("id", "unknown")
    description = feature.get("description", "no description")
    steps = feature.get("steps", [])

    logger.info(f"[generator] Starting feature {feature_id}: {description}")

    system_prompt = PROMPT_PATH.read_text()

    # Build the prompt
    progress = read_progress()
    progress_header = format_progress_header(progress) if progress else "No progress file yet.\n"

    steps_text = "\n".join(f"  - {s}" for s in steps) if steps else "  (no BDD steps defined)"

    prompt_parts = [
        progress_header,
        f"\n## Your Assignment\n\n",
        f"Implement feature **{feature_id}**: {description}\n\n",
        f"### BDD Scenario\n{steps_text}\n\n",
    ]

    if evaluator_feedback:
        prompt_parts.append(
            f"### Previous Evaluator Feedback (you must address these issues)\n\n"
            f"{evaluator_feedback}\n\n"
            f"Fix the issues identified above and ensure the feature passes evaluation.\n"
        )

    prompt_parts.append(
        "Implement this feature, write tests, verify everything passes, "
        "commit your changes, and update feature_list.json."
    )

    prompt = "".join(prompt_parts)

    options = make_options(
        system_prompt=system_prompt,
        role="generator",
        max_budget_usd=50.0,
    )

    cost_usd = 0.0
    error = None

    # 15 minute timeout per generator session — prevents infinite hangs
    # from interactive CLIs (create-next-app, shadcn init) or long npm installs
    SESSION_TIMEOUT = 900  # 15 minutes

    async def _run_session():
        nonlocal cost_usd
        async for message in query(prompt=prompt, options=options):
            if isinstance(message, ResultMessage):
                cost_usd = message.total_cost_usd or 0.0
                logger.info(f"[generator] Feature {feature_id} session complete: cost=${cost_usd:.4f}, turns={message.num_turns}")

    try:
        await asyncio.wait_for(_run_session(), timeout=SESSION_TIMEOUT)
    except asyncio.TimeoutError:
        error = f"Generator session timed out after {SESSION_TIMEOUT}s"
        logger.error(f"[generator] Feature {feature_id}: {error}")
    except Exception as e:
        error = str(e)
        logger.error(f"[generator] Feature {feature_id} failed: {error}")

    # Check if the feature was marked as passing
    updated_progress = read_progress()
    success = False
    if updated_progress:
        for f in updated_progress.items:
            if f.get("id") == feature_id and f.get("passes", False):
                success = True
                break

    return {
        "success": success,
        "cost_usd": cost_usd,
        "error": error,
    }

File: src/engine/agents/generator.ts
/**
 * Generator agent — implements one feature per session via TDD.
 *
 * Receives a single feature, runs a Claude Code session in the output
 * directory, and returns whether the feature was marked as passing.
 */

import { join } from "node:path";
import { query } from "@anthropic-ai/claude-agent-sdk";
import { makeQueryOptions } from "../client.js";
import { readProgress, formatProgressHeader } from "../progress.js";
import type { Feature, SalazarConfig } from "../../lib/types.js";

export interface GeneratorResult {
  success: boolean;
  costUsd: number;
  error: string | null;
}

export async function runGenerator(
  feature: Feature,
  config: SalazarConfig,
  outputDir: string,
  evaluatorFeedback?: string | null,
): Promise<GeneratorResult> {
  const { id: featureId, description } = feature;

  console.log(`[generator] Starting feature ${featureId}: ${description}`);

  // Build progress header
  const progress = readProgress(outputDir);
  const progressHeader = progress
    ? formatProgressHeader(progress)
    : "No progress file yet.\n";

  // Format BDD steps
  const stepsText =
    feature.steps.length > 0
      ? feature.steps.map((s) => `  - ${s}`).join("\n")
      : "  (no BDD steps defined)";

  const featureListPath = join(outputDir, "feature_list.json");

  // Assemble prompt parts
  const promptParts: string[] = [
    progressHeader,
    `\n## Your Assignment\n\n`,
    `Implement feature **${featureId}**: ${description}\n\n`,
    `### BDD Scenario\n${stepsText}\n\n`,
  ];

  if (evaluatorFeedback) {
    promptParts.push(
      `### Previous Evaluator Feedback (you must address these issues)\n\n` +
        `${evaluatorFeedback}\n\n` +
        `Fix the issues identified above and ensure the feature passes evaluation.\n`,
    );
  }

  promptParts.push(
    `Implement this feature, write tests, verify everything passes. ` +
      `All code should be written in: ${outputDir}\n` +
      `Update the feature list at: ${featureListPath} — set passes to true for feature ${featureId} only.\n` +
      `Do NOT run git commit — the orchestrator handles commits.`,
  );

  const prompt = promptParts.join("");

  // 5-minute timeout via AbortController passed into options
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 300_000);

  const options = makeQueryOptions({
    role: "generator",
    config,
    cwd: outputDir,
    maxTurns: 50,
    maxBudgetUsd: 50.0,
  });

  // Inject abort controller into options
  const optionsWithAbort = { ...options, abortController: controller };

  let costUsd = 0.0;
  let error: string | null = null;

  try {
    for await (const message of query({ prompt, options: optionsWithAbort })) {
      const msg = message as Record<string, unknown>;
      if (msg["type"] === "result") {
        if (typeof msg["total_cost_usd"] === "number") {
          costUsd = msg["total_cost_usd"];
        } else if (typeof msg["cost_usd"] === "number") {
          costUsd = msg["cost_usd"];
        }
        console.log(
          `[generator] Feature ${featureId} session complete: cost=$${costUsd.toFixed(4)}, turns=${msg["num_turns"] ?? "?"}`,
        );
      }
    }
  } catch (err) {
    if (err instanceof Error && err.name === "AbortError") {
      error = "Generator timed out after 300s";
      console.error(`[generator] Feature ${featureId}: ${error}`);
    } else {
      error = err instanceof Error ? err.message : String(err);
      console.error(`[generator] Feature ${featureId} failed: ${error}`);
    }
  } finally {
    clearTimeout(timeout);
  }

  // Check if the feature was marked as passing in feature_list.json
  const updatedProgress = readProgress(outputDir);
  let success = false;
  if (updatedProgress) {
    for (const f of updatedProgress.items) {
      if (f.id === featureId && f.passes) {
        success = true;
        break;
      }
    }
  }

  console.log(
    `[generator] Feature ${featureId} result: success=${success}, cost=$${costUsd.toFixed(4)}, error=${error ?? "none"}`,
  );

  return { success, costUsd, error };
}
